### PR TITLE
Host fairness with load monitor

### PR DESCRIPTION
**Pull request by eddyashton**
See #1462.

This adds caps to the TCP socket reads and ringbuffer message reads that the host thread executes in each UV loop iteration. Those changes are relatively minor. This also adds some code for monitoring the load on each node, currently by counting the bytes and messages read from the ringbuffer. There's a crude script to get some quick plots of this currently, so we could use it to gain insight on exceptional failures in the CI. If it proves useful we can build more metrics and automated analysis.

### CODE REVIEW

File: src/host/tcp.h [hunk 1]
```diff
@@ -594,6 +612,12 @@ namespace asynchost
         return;
       }
 
+      if (sz == UV_ENOBUFS)
+      {
+        LOG_DEBUG_FMT("TCP on_read reached allocation quota");
```
Comment (jumaffre) [inline src/host/tcp.h:617]: Minor: `LOG_TRACE_FMT()`?

File: tests/plot_node_load.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+import argparse
+import datetime
+import json
+import matplotlib.pyplot as plt
+import matplotlib.ticker as ticker
+import os
+
+# For consistency between plots we want a function from label (name) to colour.
+# We could do this programatically from the hashes to handle general values, but
+# this is sufficient and makes it simple to group similar messages with similar
```
Comment (jumaffre) [inline tests/plot_node_load.py:12]: with similar colours?

File: src/host/load_monitor.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      enclave_output_file.open("enclave_load.log", std::fstream::out);
+      enclave_counts = nlohmann::json::object();
+
+      // Register message handler for work_stats message from enclave
+      DISPATCHER_SET_MESSAGE_HANDLER(
+        bp, AdminMessage::work_stats, [this](const uint8_t* data, size_t size) {
+          auto [dumped_json] =
+            ringbuffer::read_message<AdminMessage::work_stats>(data, size);
+
+          const auto j = nlohmann::json::parse(dumped_json);
```
Comment (jumaffre) [inline src/host/load_monitor.h:41]: Should we check that this doesn't throw an exception here?

File: tests/plot_node_load.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (jumaffre) [inline tests/plot_node_load.py:1]: Could you attach a sample output of this script to this PR?

File: src/host/tcp.h (hunk 1 above)
Comment (eddyashton) [reply src/host/tcp.h:617]: I think this should be rare enough that `DEBUG` is fine. There's already a `TRACE` in `on_alloc` with more info. I think if you're in a `DEBUG` but not `TRACE` run, you want to recognise any unusual paths which are taken, which this is.

**Comment by eddyashton** [comment]
## Sample plots

### Enclave
enclave_load_bytes
enclave_load_count

### Host
host_load_bytes
host_load_count

File: tests/plot_node_load.py (hunk 4 above)
Comment (eddyashton) [reply tests/plot_node_load.py:1]: [Done](https://github.com/microsoft/CCF/pull/1480#issuecomment-672850824)

File: src/host/handle_ring_buffer.h [hunk 5]
```diff
@@ -20,7 +20,9 @@ namespace asynchost
   class HandleRingbufferImpl
   {
   private:
-    static constexpr size_t max_messages = 128;
+    // Maximum number of outbound ringbuffer messages which will be processed in
+    // a single iteration
+    static constexpr size_t max_messages = 256;
```
Comment (achamayou) [inline src/host/handle_ring_buffer.h:25]: @eddyashton I think we would prefer sizes everywhere.

File: src/host/main.cpp [hunk 6]
```diff
@@ -512,9 +515,15 @@ int main(int argc, char** argv)
...
   });
 
+  // reset the inbound-TCP processing quota each iteration
+  asynchost::ResetTCPReadQuota reset_tcp_quota;
+
   // regularly update the time given to the enclave
   asynchost::TimeUpdater time_updater(1);
 
+  // regularly record some load statistics
+  asynchost::LoadMonitor load_monitor(500, bp);
```
Comment (achamayou) [inline src/host/main.cpp:525]: 500ms ideally.

**Pull request merged**